fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -367,30 +396,141 @@ void parseKmsResponse(Reference<RESTKmsConnectorCtx> ctx,
+		// STATUS_OK is gating factor for REST request success
+		throw http_request_failed();
+	}
+
+	rapidjson::Document doc;
+	doc.Parse(resp->content.c_str());
+
+	checkResponseForError(ctx, doc);
+
+	Standalone<VectorRef<BlobMetadataDetailsRef>> result;
+
+	// Extract CipherKeyDetails
+	if (!doc.HasMember(BLOB_METADATA_DETAILS_TAG) || !doc[BLOB_METADATA_DETAILS_TAG].IsArray()) {
+		TraceEvent(SevWarn, "ParseBlobMetadataResponse_FailureMissingDetails", ctx->uid).log();
+		throw operation_failed();
+	}
+
+	for (const auto& detail : doc[BLOB_METADATA_DETAILS_TAG].GetArray()) {
+		if (!detail.IsObject()) {
+			TraceEvent(SevWarn, "ParseBlobMetadataResponse_FailureDetailsNotObject", ctx->uid)
+			    .detail("Type", detail.GetType());
+			throw operation_failed();
+		}
+
+		const bool isDomainIdPresent = detail.HasMember(BLOB_METADATA_DOMAIN_ID_TAG);
+		const bool isDomainNamePresent = detail.HasMember(BLOB_METADATA_DOMAIN_NAME_TAG);
+		const bool isBasePresent = detail.HasMember(BLOB_METADATA_BASE_LOCATION_TAG);
+		const bool isPartitionsPresent = detail.HasMember(BLOB_METADATA_PARTITIONS_TAG);
+		if (!isDomainIdPresent || !isDomainNamePresent || (!isBasePresent && !isPartitionsPresent)) {
+			TraceEvent(SevWarn, "ParseBlobMetadataResponse_MalformedDetail", ctx->uid)
+			    .detail("DomainIdPresent", isDomainIdPresent)
+			    .detail("DomainNamePresent", isDomainNamePresent)
+			    .detail("BaseLocationPresent", isBasePresent)
+			    .detail("PartitionsPresent", isPartitionsPresent);
+			throw operation_failed();
+		}
+
+		const int domainNameLen = detail[BLOB_METADATA_DOMAIN_NAME_TAG].GetStringLength();
+		std::unique_ptr<uint8_t[]> domainName = std::make_unique<uint8_t[]>(domainNameLen);
+		memcpy(domainName.get(), detail[BLOB_METADATA_DOMAIN_NAME_TAG].GetString(), domainNameLen);
+
+		std::unique_ptr<uint8_t[]> baseStr;
+		Optional<StringRef> base;
+		if (isBasePresent) {
+			const int baseLen = detail[BLOB_METADATA_BASE_LOCATION_TAG].GetStringLength();
+			baseStr = std::make_unique<uint8_t[]>(baseLen);
+			memcpy(baseStr.get(), detail[BLOB_METADATA_BASE_LOCATION_TAG].GetString(), baseLen);
+			base = StringRef(baseStr.get(), baseLen);
+		}
+
+		// just do extra memory copy for simplicity here
+		Standalone<VectorRef<StringRef>> partitions;
+		if (isPartitionsPresent) {
+			for (const auto& partition : doc[BLOB_METADATA_PARTITIONS_TAG].GetArray()) {
+				if (!partition.IsString()) {
+					TraceEvent("ParseBlobMetadataResponse_FailurePartitionNotString", ctx->uid)
+					    .detail("Type", partition.GetType());
+					throw operation_failed();
+				}
+				const int partitionLen = partition.GetStringLength();
+				std::unique_ptr<uint8_t[]> partitionStr = std::make_unique<uint8_t[]>(partitionLen);
+				memcpy(partitionStr.get(), partition.GetString(), partitionLen);
+				partitions.push_back_deep(partitions.arena(), StringRef(partitionStr.get(), partitionLen));
+			}
 		}
+
+		// Extract refresh and/or expiry interval if supplied
```

Comment:
We can extract out code to set `refresh/expireAt` in a separate method; avoid code duplication in two methods possibly.

Reply:
Yeah we should do this, but because the KmsConnector response for the encrypt ciphers also returned interval instead of timestamp, and the the EKP converted it to timestamp, that was a bigger refactor than I wanted to do here (this PR is already kind of large).

Reply:
Makes sense. Thinking more now, I think expecting the timestamp from external source is certainly a better design, will do this cleanup sometime soon. 
Thanks